### PR TITLE
feat: add Ratio Computers XML file import

### PR DESCRIPTION
**Pull request by zorcik**
## Summary

This PR adds native universal import support for Ratio Computers XML files (iX3M, iDive, etc.). Submersion can now detect and parse the `<diveSegment>` XML format generated by Ratio dive computers, seamlessly extracting header metadata, dive profiles, gas switches and computer details.

This feature was necessary because after updating the firmware in the dive computer, it became impossible to import logs directly from it. Libdivecomputer does not recognize new format.

## Changes

*   **Format Detection:** Added detection for the `<diveSegment>` XML root element in `format_detector.dart` to automatically identify Ratio XML files.
*   **Enums & Registry:** Registered `ratioXml` as a new `ImportFormat` and `ratio` as a `SourceApp`. Added a manual override option for "Ratio Computers (XML)".
*   **RatioXmlParser Implementation:** Created a dedicated parser that handles the full Ratio XML specification:
    *   **Timestamp Calculation:** Applies the custom Ratio epoch offset (2008-01-01) to correctly resolve UTC start times.
    *   **Unit Conversions:** Converts native integer units (depths in cm/dm, temperatures in deci-Celsius, pressure in 0.1 mbar) into standard internal units.
    *   **Header Data:** Extracts max/avg depths, dive mode (OC, CCR, etc.), water type, surface intervals, and gradient factors.
    *   **Profile Data (Samples):** Maps periodic samples to dive profiles, tracking NDL/TTS, ceiling depths, tank pressures, CNS, and OTU.
    *   **Gas Switches:** Detects active gas mix changes (O2/He) mid-dive and translates them into gas switch events.
    *   **Filename Parsing:** Extracts the computer model (e.g., `IX3M 2 PRO`), serial number, and dive number directly from the standard Ratio filename convention (`{model}_{serial}-dive_{number}-{date}_{time}.xml`).
*   **Tests:** Added comprehensive test suites (`ratio_xml_parser_test.dart` and `format_detector_test.dart`) covering header mapping, sample parsing, unit conversions, gas switch detection, filename extraction, and format detection edge cases.
## Test Plan

- [X] `flutter test` passes
- [X] `flutter analyze` passes
- [X] Manual testing on: Windows

### CODE REVIEW

File: lib/features/universal_import/data/parsers/ratio_xml_parser.dart [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+      final o2 = _intElement(sample, 'activeMixO2Percent');
+      final he = _intElement(sample, 'activeMixHePercent');
+      if (o2 != null && (o2 != prevO2 || he != prevHe)) {
+        if (prevO2 != null) {
+          // Gas switch occurred
+          gasSwitches.add(<String, dynamic>{
+            'timestamp': runtimeS,
+            'depth': depthM,
+            'o2': o2,
+            'he': he ?? 0,
```
[the remark this answers is not in the document]
Comment (zorcik) [reply lib/features/universal_import/data/parsers/ratio_xml_parser.dart]: gasSwitches array now emits the mapped tankIndex alongside o2 and he. This ensures the downstream ImportPipeline can successfully link gas switch events to the corresponding tank instance in the parsed payload.

File: lib/features/universal_import/data/parsers/ratio_xml_parser.dart [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+      if (tankPressure != null && tankId != null) {
+        tankPressures
+            .putIfAbsent(tankId, () => [])
+            .add(_TankReading(runtimeS: runtimeS, pressureBar: tankPressure));
+
+        point['allTankPressures'] = [
+          <String, dynamic>{
+            'tankIndex': tankId,
+            'pressure': tankPressure.toDouble(),
+          },
```
[the remark this answers is not in the document]
Comment (zorcik) [reply lib/features/universal_import/data/parsers/ratio_xml_parser.dart:330]: Ratio XML outputs raw IDs for tanks (e.g. tankId: 14), which previously caused allTankPressures to emit invalid indices that were ignored downstream. I added a tankIdToIndex mapping that dynamically assigns sequential 0-based index numbers as new tanks are encountered in the samples array. Now, allTankPressures correctly references a sequential tankIndex (0, 1, 2, etc.) matching the layout of the diveData['tanks'] list.

**Comment by ericgriffin** [comment]
Thanks for this. The tank index and gas switch rework both look good to me.

Three things still outstanding (one of them blocking).

#### 1. Surface pressure is 1000x too large (blocking)

`lib/features/universal_import/data/parsers/ratio_xml_parser.dart:188`

```dart
diveData['surfacePressure'] = surfacePressure / 10.0;  // yields 997.1
```

The canonical unit for this key is **bar**, not mbar. That's documented at both the entity and the schema:

- `lib/features/dive_log/domain/entities/dive.dart:85` -> `final double? surfacePressure; // bar (standard ~1.013)`
- `lib/core/database/database.dart:647` -> `RealColumn get surfacePressure => real().nullable()(); // bar (default ~1.013)`

For reference, the other import paths all normalize to bar: `uddf_import_service.dart:251` does `pascal / 100000`, `uddf_export_builders.dart:170` multiplies back by `100000`, and `shearwater_dive_mapper.dart:315` calls `mbarToBar` explicitly.

Since `<surfacePressureMbar>` is in tenths of a mbar, the conversion is:

```dart
diveData['surfacePressure'] = surfacePressure / 10000.0;  // 9971 -> 0.9971 bar
```

The test at `test/features/universal_import/data/parsers/ratio_xml_parser_test.dart:222` currently asserts `closeTo(997.1, 0.1)`, which is why this stayed green. It needs to become `closeTo(0.9971, 0.0001)`.

#### 2. `desaturationTime` is dead data

`ratio_xml_parser.dart:192-194` parses `<desaturationTimeS>` into `diveData['desaturationTime']`, but there is no such column, entity field, or importer branch anywhere in the codebase. Nothing consumes it. Please drop it rather than leaving a key that looks supported but isn't. If you think it's worth persisting, that's a separate PR with a schema change.

#### 3. Payload metadata key doesn't match the convention

`ratio_xml_parser.dart:129`

```dart
metadata: {
  'sourceApp': 'Ratio Computers',
  ...
}
```

Every other parser uses a snake_case `source` key: `{'source': 'macdive_xml'}`, `{'source': 'dan_dl7'}`, `{'source': 'shearwater_cloud'}`, `{'source': 'subsurface_xml'}`. Please use `{'source': 'ratio_xml'}` for consistency. Related: the early-return error payloads in `parse()` omit `metadata` entirely, where the other parsers still stamp `source` on their error paths (see `macdive_xml_parser.dart:50`, `:66`, `:82`).

#### 4. The two fixes aren't covered by tests

This one isn't blocking, but I'd rather not merge the fixes untested.

The whole test file changed by one line (`tankIndex` 14 -> 0). The gas switch test at `ratio_xml_parser_test.dart:414` still asserts only `timestamp`, `o2`, `he`, and `depth`. It never asserts `tankIndex`, which is the single field that decides whether the switch survives import: `uddf_entity_importer.dart:1690-1708` resolves a switch via `tankRef`, then `gasMixRef`, then `tankIndex`, and returns `null` if all three miss. As written, someone could revert the `switchEvent['tankIndex'] = currentTankIndex` line at `ratio_xml_parser.dart:305` and the suite would stay green.

Could you add:

- `expect(switches[0]['tankIndex'], 1)` to the existing two-tank gas switch test.
- A multi-tank case asserting `tanks` has length 2 and that each entry carries the right `gasMix`, covering the new `tankMixes` derivation.

#### One broader note

Every fixture in the test file is built by the synthetic `_buildRatioXml` helper, so assumptions like "tankPressure is in bar" and "depthMax is in cm" are only checked against a fixture that already assumes them. That's how the surface pressure unit slipped through. If you can attach one real iX3M export as a fixture (see `test/core/services/export/uddf/uddf_macdive_real_sample_test.dart` for the pattern we use elsewhere), it would firm up the unit assumptions considerably. Not a merge blocker, but worth doing while the format is fresh in your head.

One note on item 1 before you post: I derived / 10000.0 from the PR's own stated encoding (tenths of a mbar) combined with the documented bar unit. If a real iX3M file turns out to write plain mbar in that field rather than tenths, the divisor is 1000.0 and the expected value is 9.971 bar, which would be obviously wrong and caught immediately. Worth confirming against a real export, which is part of why item 4's fixture request matters.

**Comment by zorcik** [comment]
Thanks for the detailed review! I've gone ahead and pushed updates to address all these points.

**Surface Pressure Unit (Fixed)**: You were absolutely right. The unit conversion has been updated to divide the raw `surfacePressureMbar` by 10000.0, safely bringing the tenths-of-a-mbar value into canonical bar format (0.9971 bar). The unit test assertion has also been updated to strictly expect this float value.

**Desaturation Time (Removed):** The desaturationTimeS parsing block has been completely dropped from the parser to avoid polluting the payload with unmapped, dead keys.

**Metadata Convention (Fixed):** The metadata key has been updated from `sourceApp: 'Ratio Computers'` to `source: 'ratio_xml'`. Additionally, I went ahead and added metadata: `const {'source': 'ratio_xml'}` to all four of the early-return error paths in `parse()`, so it should match the `macdive_xml_parser.dart` standard behavior.

**Gas Switch Test Coverage (Added):** The existing gas switch test has been augmented to strictly assert that` switches[0]['tankIndex'] == 1`. In the same block, it also asserts that tanks generates precisely 2 entries with the exact synthesized gasMix (21% and 50% O2 respectively), comprehensively locking in the new fallback behavior.

**Integration Fixture (Added):** I've anonymized the real iX3M 2 PRO XML export and committed it to `test/fixtures/universal_import/ratio_xml/`. It is now directly wired up to a new synthetic regression suite (`ratio_xml_real_sample_test.dart`) that explicitly asserts the extracted max depth, avg depth, timestamp, and surface pressure exactly match their real-world intended values, heavily firming up all of our unit assumptions.

Everything should be fully covered by tests. Flutter test results are quite hard to read for me (I've got many errors from test that should be working).

Let me know if there's anything else you need before merging!

**Review by ericgriffin — APPROVED** [review]
👍
